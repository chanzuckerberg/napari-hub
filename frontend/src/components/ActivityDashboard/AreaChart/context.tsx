import { createContext, ReactNode, useContext, useMemo } from 'react';

interface AreaChartStateValue {
  publicReleaseX: number;
  isScreen600: boolean;
  visibleMonths: number[];
}

const AreaChartStateContext = createContext<AreaChartStateValue>({
  publicReleaseX: 0,
  isScreen600: false,
  visibleMonths: [],
});

interface Props extends AreaChartStateValue {
  children: ReactNode;
}

/**
 * Provider for sharing state used in the area chart component.
 */
export function AreaChartStateProvider({
  children,
  isScreen600,
  publicReleaseX,
  visibleMonths,
}: Props) {
  const value = useMemo<AreaChartStateValue>(
    () => ({
      publicReleaseX,
      isScreen600,
      visibleMonths,
    }),
    [isScreen600, publicReleaseX, visibleMonths],
  );

  return (
    <AreaChartStateContext.Provider value={value}>
      {children}
    </AreaChartStateContext.Provider>
  );
}

export function useAreaChartState(): AreaChartStateValue {
  return useContext(AreaChartStateContext);
}
