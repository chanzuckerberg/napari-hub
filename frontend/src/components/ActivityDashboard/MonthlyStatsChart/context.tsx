import dayjs from 'dayjs';
import { createContext, ReactNode, useContext, useMemo } from 'react';

import { I18nKeys } from '@/types/i18n';

interface MonthlyStatsValue {
  dateLineI18nKey: I18nKeys<'activity'>;
  startDate: dayjs.ConfigType;
}

const MonthlyStatsContext = createContext<MonthlyStatsValue | null>(null);

interface Props
  extends Pick<MonthlyStatsValue, 'startDate' | 'dateLineI18nKey'> {
  children: ReactNode;
}

export function MonthlyStatsProvider({
  children,
  dateLineI18nKey,
  startDate,
}: Props) {
  const value = useMemo(
    () => ({
      dateLineI18nKey,
      startDate,
    }),
    [dateLineI18nKey, startDate],
  );

  return (
    <MonthlyStatsContext.Provider value={value}>
      {children}
    </MonthlyStatsContext.Provider>
  );
}

export function useMonthlyStats(): MonthlyStatsValue {
  const value = useContext(MonthlyStatsContext);

  if (!value) {
    throw new Error('useMonthlyStats() must be used in a MonthlyStatsProvider');
  }

  return value;
}
