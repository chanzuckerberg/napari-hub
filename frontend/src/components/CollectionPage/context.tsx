import { createContext, ReactNode, useContext } from 'react';

import { CollectionData } from '@/types/collections';

const CollectionContext = createContext<CollectionData | null>(null);

interface Props {
  collection: CollectionData;
  children: ReactNode;
}

/**
 * Provider for access to collection state from anywhere in collection page.
 */
export function CollectionContextProvider({ collection, children }: Props) {
  return (
    <CollectionContext.Provider value={collection}>
      {children}
    </CollectionContext.Provider>
  );
}

/**
 * Hook for accessing collection data from hub API.
 */
export function useCollection(): CollectionData {
  const collection = useContext(CollectionContext);

  if (!collection) {
    throw new Error(
      'useCollections() may not be used outside of CollectionContextProvider',
    );
  }

  return collection;
}
