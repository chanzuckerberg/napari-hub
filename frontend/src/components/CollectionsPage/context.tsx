import { createContext, ReactNode, useContext } from 'react';

import { CollectionIndexData } from '@/types/collections';

const CollectionsContext = createContext<CollectionIndexData[]>([]);

interface Props {
  collections: CollectionIndexData[];
  children: ReactNode;
}

/**
 * Provider for access to collections state from anywhere in collections home
 * page UI.
 */
export function CollectionsContextProvider({ collections, children }: Props) {
  return (
    <CollectionsContext.Provider value={collections}>
      {children}
    </CollectionsContext.Provider>
  );
}

/**
 * Hook for accessing list of collections from hub API.
 */
export function useCollections(): CollectionIndexData[] {
  return useContext(CollectionsContext);
}
