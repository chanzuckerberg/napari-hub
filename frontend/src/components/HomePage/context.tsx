import { createContext, ReactNode, useContext, useMemo } from 'react';

import { PluginSectionsResponse } from '@/types';

interface HomePageContextValue {
  pluginSections: PluginSectionsResponse;
}

const HomePageContext = createContext<HomePageContextValue | null>(null);

interface Props extends Pick<HomePageContextValue, 'pluginSections'> {
  children: ReactNode;
}

export function HomePageProvider({ children, pluginSections }: Props) {
  const value = useMemo(
    () => ({
      pluginSections,
    }),
    [pluginSections],
  );

  return (
    <HomePageContext.Provider value={value}>
      {children}
    </HomePageContext.Provider>
  );
}

export function useHomePage(): HomePageContextValue {
  const value = useContext(HomePageContext);

  if (!value) {
    throw new Error('useHomePage must be used in a HomePageProvider');
  }

  return value;
}
