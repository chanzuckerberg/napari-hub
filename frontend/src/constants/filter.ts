import {
  filterLinuxState,
  filterMacState,
  filterPython37State,
  filterPython38State,
  filterPython39State,
  filterWindowsState,
} from '@/store/search/filter.state';
import { StateLabelEntry } from '@/utils';

export const PYTHON_LABEL_ENTRIES: StateLabelEntry[] = [
  {
    label: '3.7',
    state: [filterPython37State],
  },
  {
    label: '3.8',
    state: [filterPython38State],
  },
  {
    label: '3.9',
    state: [filterPython39State],
  },
];

export const OPERATING_SYSTEM_LABEL_ENTRIES: StateLabelEntry[] = [
  {
    label: 'Linux',
    state: filterLinuxState,
  },
  {
    label: 'macOS',
    state: filterMacState,
  },
  {
    label: 'Windows',
    state: filterWindowsState,
  },
];
