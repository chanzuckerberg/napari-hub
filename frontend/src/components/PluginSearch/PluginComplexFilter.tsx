import styled from '@emotion/styled';
import Popper from '@material-ui/core/Popper';
import { AutocompleteRenderOptionState } from '@material-ui/lab/Autocomplete';
import clsx from 'clsx';
import {
  Button,
  ComplexFilter,
  DefaultMenuSelectOption,
  InputDropdownProps,
} from 'czifui';
import { get, isEqual } from 'lodash';
import { useTranslation } from 'next-i18next';
import { useCallback, useEffect, useRef, useState } from 'react';
import { render } from 'react-dom';
import { subscribe, useSnapshot } from 'valtio';

import {
  CheckboxIcon,
  ChevronDown,
  ChevronUp,
  Search,
} from '@/components/common/icons';
import { useSearchStore } from '@/store/search/context';
import { FilterKey } from '@/store/search/search.store';
import { appTheme } from '@/theme';

import styles from './PluginComplexFilter.module.scss';
import { useFilterLabels } from './useFilterLabels';
import { useFilterOptionLabels } from './useFilterOptionLabels';
import { useWorkflowStepGroups } from './useWorkflowStepGroups';

const SEARCH_ENABLED_FILTERS = new Set<FilterKey>(['workflowStep']);

const CATEGORY_FILTERS = new Set<FilterKey>(['workflowStep', 'imageModality']);

interface Props {
  filterKey: FilterKey;
}

/**
 * Option data structure used in the complex filter autocomplete component.
 */
interface PluginMenuSelectOption extends DefaultMenuSelectOption {
  /**
   * Key used for identifying what state to use when the option is enabled / disabled.
   */
  stateKey: string;
}

const StyledPopper = styled(Popper)`
  background: #fff;
  box-shadow: ${appTheme.shadows?.[10]};
  display: flex;
  flex-direction: column;
  max-width: 225px;
  padding: 0.75rem 0;
  width: 100%;

  .MuiAutocomplete-popper {
    background: #fff;
    position: relative;
    width: 100% !important;
  }

  .MuiAutocomplete-option {
    min-height: 0;
  }
`;

function InputDropdown(props: InputDropdownProps) {
  const { label, sdsStage } = props;
  const iconSizeClassName = 'w-4 h-4';

  return (
    <Button
      {...props}
      className="py-2 px-0 text-black w-full flex justify-between"
      classes={{
        label: 'border-b border-black pb-2',
      }}
    >
      <span className="font-semibold text-sm">{label}</span>
      {sdsStage === 'default' && <ChevronDown className={iconSizeClassName} />}
      {sdsStage === 'userInput' && <ChevronUp className={iconSizeClassName} />}
    </Button>
  );
}

/**
 * Complex filter component for filtering a specific part of the filter state.
 */
export function PluginComplexFilter({ filterKey }: Props) {
  const [t] = useTranslation(['common', 'pageTitles', 'pluginData']);
  const workflowStepGroups = useWorkflowStepGroups();
  const { searchStore } = useSearchStore();
  const state = useSnapshot(searchStore);

  // Get dictionary that maps category keys to category names in the user's
  // language.
  const categoryNamesMap = t('pluginData:category.labels') as Record<
    string,
    string | undefined
  >;

  const filterStore = searchStore.filters[filterKey];
  const filterState = state.filters[filterKey];

  const filterOptionLabels = useFilterOptionLabels();
  const filterLabels = useFilterLabels();

  // Store options in ref to prevent re-render on options value. There's a race
  // condition issue when re-rendering ComplexFilter when `options` and
  // `pendingValue` are updated at the same time.
  const optionsRef = useRef(
    Object.keys(filterState).map((stateKey) => {
      const optionLabel = filterOptionLabels[stateKey];
      const categoryName = categoryNamesMap[stateKey];

      let name = stateKey;

      if (optionLabel) {
        name = optionLabel;
      } else if (categoryName) {
        name = categoryName;
      }

      return {
        name,
        stateKey,
      };
    }),
  );

  const getEnabledOptions = useCallback(
    () =>
      optionsRef.current.filter((option) =>
        get(filterStore, option.stateKey, false),
      ),
    [filterStore],
  );

  const [pendingState, setPendingState] = useState<PluginMenuSelectOption[]>(
    // Use truthy values in filter state as the initial value. Options need to
    // be taken from `optionsRef` because the autocomplete component compares
    // values using referential equality.
    getEnabledOptions,
  );

  // Effect for keeping the local pending state in sync with external changes
  // to the filter store.
  useEffect(
    () =>
      subscribe(filterStore, () => {
        const options = getEnabledOptions();

        // Only update the state if it changed to prevent unnecessary re-renders.
        if (!isEqual(options, pendingState)) {
          setPendingState(options);
        }
      }),
    [filterStore, getEnabledOptions, pendingState],
  );

  // Effect that merges the pending state into the global state.
  useEffect(() => {
    const enabledStates = new Set(
      pendingState.map((option) => option.stateKey),
    );
    const nextState: Record<string, boolean> = {};

    for (const key of Object.keys(filterStore)) {
      nextState[key] = enabledStates.has(key);
    }

    Object.assign(filterStore, nextState);
  }, [filterStore, pendingState]);

  const isSearchEnabled = SEARCH_ENABLED_FILTERS.has(filterKey);

  // Effect that sets the max width of the tooltip container when it is added to
  // the DOM. This is required so that the width is only as wide as the filter
  // component. Otherwise, it would fill the whole screen.
  useEffect(() => {
    const observer = new MutationObserver((mutations) =>
      mutations.forEach((mutation) =>
        // Only observe added nodes.
        mutation.addedNodes.forEach((node) => {
          const tooltip = node as HTMLDivElement;
          // Check if current added node is a tooltip
          if (tooltip.getAttribute('role') === 'tooltip') {
            // Flag for if the tooltip is for the current filter.
            const isActiveFilter = !!tooltip.querySelector(
              `[data-filter=${filterKey}]`,
            );

            // Get the filter node for the current filter.
            const complexFilterNode =
              isActiveFilter &&
              document.querySelector(
                `[data-complex-filter][data-filter=${filterKey}]`,
              );

            if (complexFilterNode) {
              const width = complexFilterNode.clientWidth;

              // Set the width of the tooltip to be the same width as the filter
              // component. If the screen width is smaller than 300px, then
              // default to 100% because the tooltip does not fill the screen
              // for sizes <300px.
              tooltip.style.maxWidth =
                window.outerWidth < 300 ? '100%' : `${width}px`;

              // Manually set the placeholder of the input component. This
              // workaround is required until props are added to the
              // ComplexFilter component that'll let us pass props to the nested
              // Autocomplete input component.
              const searchInput = tooltip.querySelector('input');
              if (searchInput) {
                searchInput.placeholder = 'search for a category';
              }

              // Manually replace the filter input adornment. This workaround is
              // required for the same reasons as above.
              if (filterKey === 'workflowStep') {
                const svgContainer = tooltip.querySelector(
                  '.MuiInputAdornment-root',
                );

                if (svgContainer) {
                  render(<Search />, svgContainer);
                }
              }
            }
          }
        }),
      ),
    );

    observer.observe(document.body, { childList: true });

    return () => observer.disconnect();
  }, [filterKey]);

  return (
    <ComplexFilter
      className={clsx(
        styles.complexFilter,
        CATEGORY_FILTERS.has(filterKey) && styles.categories,
      )}
      data-testid="pluginFilter"
      // Data attribute used to query the complex filter node.
      data-complex-filter
      // Data attribute used to query the complex filter node by filter.
      data-filter={filterKey}
      label={filterLabels[filterKey]}
      multiple
      search={isSearchEnabled}
      onChange={(nextOptions) =>
        setPendingState(nextOptions as PluginMenuSelectOption[])
      }
      InputDropdownComponent={InputDropdown}
      MenuSelectProps={{
        classes: {
          groupLabel:
            'text-[0.875rem] leading-normal text-black font-semibold top-0',
          root: clsx('px-4', styles.autoComplete, isSearchEnabled && 'mb-3'),
          noOptions: 'text-[0.875rem]',
        },

        noOptionsText: 'No categories',

        // Data attribute used for querying the tooltip for the current filter.
        'data-filter': filterKey,

        groupBy: (option: PluginMenuSelectOption) =>
          workflowStepGroups[option.stateKey] ?? '',

        renderOption: (
          option: PluginMenuSelectOption,
          { selected }: AutocompleteRenderOptionState,
        ) => (
          <div className="flex space-x-2 py-1 px-4">
            <CheckboxIcon
              className="min-w-[0.875rem] min-h-[0.875rem]"
              checked={selected}
            />

            <p
              className={clsx(
                '-mt-1 text-[0.875rem] break-words leading-normal',
                selected && 'text-black',
              )}
            >
              {option.name}
            </p>
          </div>
        ),
      }}
      options={optionsRef.current}
      value={pendingState}
      PopperComponent={StyledPopper}
    />
  );
}
