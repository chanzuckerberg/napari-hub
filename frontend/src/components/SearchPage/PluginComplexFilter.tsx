import styled from '@emotion/styled';
import { AutocompleteRenderOptionState } from '@mui/material/Autocomplete';
import IconButton from '@mui/material/IconButton';
import clsx from 'clsx';
import {
  Button,
  ComplexFilter,
  DefaultMenuSelectOption,
  DropdownPopper,
  InputDropdownProps,
  Tooltip,
} from 'czifui';
import { get, isEqual } from 'lodash';
import { useTranslation } from 'next-i18next';
import {
  ComponentProps,
  useCallback,
  useEffect,
  useRef,
  useState,
} from 'react';
import { subscribe, useSnapshot } from 'valtio';

import { CheckboxIcon, ChevronDown, ChevronUp, Info } from '@/components/icons';
import { useSearchStore } from '@/store/search/context';
import { FilterKey } from '@/store/search/search.store';
import { theme } from '@/theme';

import styles from './PluginComplexFilter.module.scss';
import { useFilterLabels } from './useFilterLabels';
import { useFilterOptionLabels } from './useFilterOptionLabels';
import { useWorkflowStepGroups } from './useWorkflowStepGroups';

/**
 * Metadata filters that should have a search bar.
 */
const SEARCH_ENABLED_FILTERS = new Set<FilterKey>(['workflowStep', 'authors']);

const CATEGORY_FILTERS = new Set<FilterKey>(['workflowStep', 'imageModality']);

interface Props {
  filterKey: FilterKey;
}

/**
 * Option data structure used in the complex filter autocomplete component.
 */
interface PluginMenuSelectOption extends DefaultMenuSelectOption {
  /**
   * Key used for identifying what state to use when the option is enabled / disabled.
   */
  stateKey: string;

  /**
   * Tooltip to render with option for describing its usage.
   */
  tooltip?: string;
}

function Popper({
  className,
  ...props
}: ComponentProps<typeof DropdownPopper>) {
  return (
    <DropdownPopper className={clsx(className, styles.popper)} {...props} />
  );
}

const StyledPopper = styled(Popper)`
  background: #fff;
  box-shadow: ${theme.shadows[10]};
  display: flex;
  flex-direction: column;
  padding: 0.75rem 0;
  width: 100%;

  max-width: calc(100vw - 50px - 8px);

  @media (min-width: 495px) {
    max-width: calc(100vw - 100px - 16px);
  }

  @media (min-width: 875px) {
    max-width: calc(225px);
  }

  .MuiAutocomplete-popper {
    background: #fff;
    position: relative;
    width: 100% !important;
  }

  .MuiAutocomplete-option {
    min-height: 0;
  }
`;

function InputDropdown(props: InputDropdownProps) {
  const { label, sdsStage } = props;
  const iconSizeClassName = 'w-4 h-4';

  return (
    <div className="border-b border-black">
      <Button
        {...props}
        className="text-black w-full flex justify-between p-0 pb-2"
      >
        <span className="font-semibold text-sm">{label}</span>
        {sdsStage === 'default' && (
          <ChevronDown className={iconSizeClassName} />
        )}
        {sdsStage === 'userInput' && (
          <ChevronUp className={iconSizeClassName} />
        )}
      </Button>
    </div>
  );
}

const getFirstLetter = (value: string) => value[0].toUpperCase();

/**
 * Complex filter component for filtering a specific part of the filter state.
 */
export function PluginComplexFilter({ filterKey }: Props) {
  const [t] = useTranslation(['common', 'pageTitles', 'pluginData']);
  const workflowStepGroups = useWorkflowStepGroups();
  const { searchStore } = useSearchStore();
  const state = useSnapshot(searchStore);

  // Get dictionary that maps category keys to category names in the user's
  // language.
  const categoryNamesMap = t('pluginData:category.labels') as Record<
    string,
    string | undefined
  >;

  const filterStore = searchStore.filters[filterKey];
  const filterState = state.filters[filterKey];

  const filterOptionLabels = useFilterOptionLabels();
  const filterLabels = useFilterLabels();

  // Store options in ref to prevent re-render on options value. There's a race
  // condition issue when re-rendering ComplexFilter when `options` and
  // `pendingValue` are updated at the same time.
  const optionsRef = useRef(
    Object.keys(filterState)
      .map((stateKey) => {
        const optionLabel = filterOptionLabels[stateKey];
        const categoryName = categoryNamesMap[stateKey];

        let name = stateKey;
        let tooltip: string | undefined;

        if (optionLabel) {
          name = optionLabel.label;
          tooltip = optionLabel.tooltip;
        } else if (categoryName) {
          name = categoryName;
        }

        return {
          name,
          stateKey,
          tooltip,
        };
      })
      .sort((option1, option2) => {
        switch (filterKey) {
          // Sort authors alphabetically.
          case 'authors': {
            return option1.name.localeCompare(option2.name);
          }

          default:
            return 0;
        }
      }),
  );

  const getEnabledOptions = useCallback(
    () =>
      optionsRef.current.filter((option) =>
        get(filterStore, option.stateKey, false),
      ),
    [filterStore],
  );

  const [pendingState, setPendingState] = useState<PluginMenuSelectOption[]>(
    // Use truthy values in filter state as the initial value. Options need to
    // be taken from `optionsRef` because the autocomplete component compares
    // values using referential equality.
    getEnabledOptions,
  );

  // Effect for keeping the local pending state in sync with external changes
  // to the filter store.
  useEffect(
    () =>
      subscribe(filterStore, () => {
        const options = getEnabledOptions();

        // Only update the state if it changed to prevent unnecessary re-renders.
        if (!isEqual(options, pendingState)) {
          setPendingState(options);
        }
      }),
    [filterStore, getEnabledOptions, pendingState],
  );

  // Effect that merges the pending state into the global state.
  useEffect(() => {
    const enabledStates = new Set(
      pendingState.map((option) => option.stateKey),
    );
    const nextState: Record<string, boolean> = {};

    for (const key of Object.keys(filterStore)) {
      nextState[key] = enabledStates.has(key);
    }

    Object.assign(filterStore, nextState);
  }, [filterStore, pendingState]);

  const isSearchEnabled = SEARCH_ENABLED_FILTERS.has(filterKey);

  return (
    <ComplexFilter
      className={clsx(
        'py-2 hover:bg-hub-gray-100',
        styles.complexFilter,

        CATEGORY_FILTERS.has(filterKey) && styles.categories,
      )}
      data-testid="pluginFilter"
      // Data attribute used to query the complex filter node.
      data-complex-filter
      // Data attribute used to query the complex filter node by filter.
      data-filter={filterKey}
      label={filterLabels[filterKey]}
      multiple
      search={isSearchEnabled}
      onChange={(nextOptions) =>
        setPendingState(nextOptions as PluginMenuSelectOption[])
      }
      InputDropdownComponent={InputDropdown}
      DropdownMenuProps={{
        PopperBaseProps: {
          // className: 'bg-red-100',
        },

        classes: {
          groupLabel: clsx(
            '!text-sm leading-normal text-black font-semibold top-0 px-2',

            !(filterKey === 'workflowStep' || filterKey === 'authors') &&
              'hidden',
          ),

          root: clsx(styles.autoComplete, isSearchEnabled && 'mb-sds-l'),

          noOptions: 'text-sm',
        },

        noOptionsText: 'No categories',

        // Data attribute used for querying the tooltip for the current filter.
        'data-filter': filterKey,

        groupBy: (option: PluginMenuSelectOption) => {
          switch (filterKey) {
            case 'workflowStep':
              return workflowStepGroups[option.stateKey] ?? '';

            case 'authors':
              // Use first letter of name for author grouping.
              return getFirstLetter(option.name);

            default:
              return null;
          }
        },

        renderOption: (
          optionProps: Record<string, unknown>,
          option: PluginMenuSelectOption,
          { selected }: AutocompleteRenderOptionState,
        ) => (
          <div
            {...optionProps}
            className={clsx(
              'flex flex-auto justify-between p-2',
              'cursor-pointer hover:bg-hub-gray-100',
            )}
          >
            <div className="flex space-x-sds-s">
              <CheckboxIcon
                className="min-w-[0.875rem] min-h-[0.875rem]"
                checked={selected}
              />

              <p
                className={clsx(
                  '-mt-sds-xxs text-[0.875rem] break-words leading-normal',
                  selected && 'text-black',
                )}
              >
                {option.name}
              </p>
            </div>

            {option.tooltip && (
              <Tooltip
                disableInteractive
                leaveDelay={0}
                title={
                  <div>
                    <p className="font-semibold text-[0.6875rem] screen-495:text-xs">
                      {option.name}
                    </p>

                    <span className="text-[0.5625rem] screen-495:text-[0.6875rem]">
                      {option.tooltip}
                    </span>
                  </div>
                }
              >
                <IconButton>
                  <Info />
                </IconButton>
              </Tooltip>
            )}
          </div>
        ),
      }}
      options={optionsRef.current.filter(
        (option) => option.name && option.stateKey,
      )}
      value={pendingState}
      PopperComponent={StyledPopper}
    />
  );
}
