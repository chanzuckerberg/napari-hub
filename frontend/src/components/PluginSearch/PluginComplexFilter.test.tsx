import {
  fireEvent,
  getByText as getByTextFromContainer,
  render,
  screen,
} from '@testing-library/react';
import { get } from 'lodash';

import homePageI18n from '@/i18n/en/homePage.json';
import pluginDataI18n from '@/i18n/en/pluginPage.json';
import { SearchStoreProvider } from '@/store/search/context';
import { SearchFilterStore } from '@/store/search/filter.store';
import { PluginSearchStore } from '@/store/search/search.store';

import { PluginComplexFilter } from './PluginComplexFilter';

jest.mock('next-i18next', () => ({
  useTranslation: () => [
    (i18nKey: string) => {
      const key = i18nKey.split(':')[1];
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return get(homePageI18n, key) || get(pluginDataI18n, key) || i18nKey;
    },
  ],
}));

describe('<PluginComplexFilter />', () => {
  let mockSearchStore: PluginSearchStore;

  function setOsFilterState(
    state: Partial<SearchFilterStore['operatingSystems']>,
  ) {
    Object.assign(mockSearchStore.filters.operatingSystems, state);
  }

  function renderFilter() {
    return render(
      <SearchStoreProvider searchStore={mockSearchStore}>
        <PluginComplexFilter filterKey="operatingSystems" />
      </SearchStoreProvider>,
    );
  }

  beforeEach(() => {
    mockSearchStore = new PluginSearchStore();
  });

  it('should match snapshot', () => {
    setOsFilterState({
      linux: true,
      mac: true,
      windows: true,
    });

    const component = renderFilter();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render filter options on initial load', () => {
    setOsFilterState({
      linux: true,
      windows: true,
    });

    const { queryByText } = renderFilter();

    expect(queryByText('Linux')).toBeInTheDocument();
    expect(queryByText('Windows')).toBeInTheDocument();
    expect(queryByText('MacOS')).toBeNull();
  });

  it('should filter when menu item is clicked', () => {
    const { getByTestId } = renderFilter();

    expect(screen.queryByTestId('Windows')).toBeNull();
    expect(screen.queryByTestId('Linux')).toBeNull();

    const pluginFilter = getByTestId('pluginFilter');
    const pluginFilterButton = pluginFilter.querySelector('button');

    // Open menu
    if (pluginFilterButton) {
      fireEvent.click(pluginFilterButton);
    }

    // Click on menu options
    for (const option of ['Windows', 'Linux']) {
      const optionElement = getByTextFromContainer(
        screen.getByRole('tooltip'),
        option,
      );
      fireEvent.click(optionElement);
    }

    expect(screen.getAllByText('Windows')[0]).toBeInTheDocument();
    expect(screen.getAllByText('Linux')[0]).toBeInTheDocument();
  });

  it('should remove filters', () => {
    setOsFilterState({
      linux: true,
      windows: true,
    });

    const { getByText, queryByText } = renderFilter();

    // Click delete icon on the chip.
    const linuxChip = getByText('Linux');
    const deleteButton = linuxChip.parentElement?.querySelector(
      '.MuiChip-deleteIcon',
    );

    if (deleteButton) {
      fireEvent.click(deleteButton);
    }

    expect(queryByText('Linux')).toBeNull();
  });
});
