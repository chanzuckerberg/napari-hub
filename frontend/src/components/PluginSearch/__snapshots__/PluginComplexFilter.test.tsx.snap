// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PluginComplexFilter /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-end8kr"
    data-testid="pluginFilter"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text css-gdpnou"
      label="Operating system"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span>
          Operating system
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          fill="none"
          fillcontrast="white"
          focusable="false"
          height="14"
          viewBox="0 0 14 14"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M6.973 8.977l-5.19-5.328-1.14 1.078 6.328 6.49 6.375-6.423-.886-.963-.184-.081h-.055l-.178.075-5.07 5.152z"
            fill="#3867FA"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div>
      <div
        class="MuiChip-root css-1mixguo MuiChip-deletable"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          Linux
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
      <div
        class="MuiChip-root css-1mixguo MuiChip-deletable"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          macOS
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
      <div
        class="MuiChip-root css-1mixguo MuiChip-deletable"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiChip-label"
        >
          Windows
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiChip-deleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
